lib.types: Add deferredModule

roberth opened the pull request:
###### Description of changes

Great for submodule defaults. Example: `network.defaults` in NixOps.

Accepts modules, returns a list of modules, with a good default module location.

 I'll quote the docs below. Also look at the example.

`types.deferredModule`

Whereas `submodule` represents an option tree, `deferredModule` represents a module value, such as a module file or a configuration.

It can be set multiple times.

Module authors can use its value, which is always a list of module values, in `imports` or in `submoduleWith`'s `modules` parameter.
Note that `imports` must be evaluated before the module fixpoint. Because of this, deferred modules can only be imported into "other" fixpoints, such as submodules.

One use case for this type is the type of a "default" module that allow the user to affect all submodules in an `attrsOf submodule` at once. This is more convenient and discoverable than expecting the module user to type-merge with the `attrsOf submodule` option.

Refs https://github.com/NixOS/nixops/pull/1508

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

roberth committed:
lib.types: Add deferredModule

roberth commented:
cc @infinisil

roberth committed:
lib/test/modules.sh: Test deferredModule error location file

roberth committed:
lib.types.deferredModule: Improve reported location

infinisil commented on lib/types.nix:
```diff
@@ -537,6 +537,14 @@ rec {
+    deferredModule = mkOptionType {
+      name = "deferredModule";
+      description = "module";
+      check = t: isAttrs t || isFunction t;
```
Shouldn't this also allow paths?

infinisil replied:
And how about also allowing lists of modules, instead of just a single one

infinisil replied:
Hmm, it bothers me a bit that the result of this type is a list, but the input is a single element. This is in contrast to essentially all other types whose input and output types are generally the same.

Also, if the input type is also a list (which I think also makes more sense, since it's how `imports`, the `modules` argument for `evalModules`/`submoduleWith` works), then it might make sense to use the `listOf` type to implement it.

Which then brings me to my general dislike of lists, the main problem being that it's not possible to remove entries once added. I guess in this case it's somewhat fine though, because I believe `disabledModules` can be used to essentially remove modules from the list, though finding out how to actually refer to modules in the list is pretty hard.

roberth replied:
This reminds me of [`Html` in Elm](https://discourse.elm-lang.org/t/why-is-html-a-single-dom-node/3966/2), which is also an inherently monoidal type that they're constantly putting in lists (the free monoid) for no good reason. Similar to their `[Html] -> Html` operation (`<div>` or `<span>`), our concat operation (`x: { imports = x; }`) is also leaky. It leaks through in error messages with `anon-`-something-something strings.

I've embraced the leakiness and I've work with the rules:
 1. keep the conversions to a minimum (they're leaky)
 2. keep the interface simple
     - we don't have a use case for defining lists of modules, and it's still covered by `mkMerge`
     - `submodule` doesn't support lists, so why should `deferredModule`?

It'd be a different store if lists of modules were _always_ considered to be valid modules. This would include the `submodule` argument, `submodule` definitions, `imports` _items_, `evalModules { modules }` items, etc.


> my general dislike of lists

I share the dislike, as they imply ordering, which is often not necessary or desirable in a declarative system. In this case though, they are only as unnecessary and undesirable as in the rest of the modules system - not something we can expect to solve here.



> `disabledModules`

The best solution to removals is not to include the things in the first place.
`disabledModules` should operate just fine on the parent module and on path-imported ones. Anonymous ones will be equally problematic as anonymous submodules.


> though finding out how to actually refer to modules in the list is pretty hard.

Do you mean list indexing? That's not something I'd think of, but I guess someone could try - and eventually fail horribly.

I guess this might be reason to make it slightly harder? Although `x.imports !! y` isn't much of an obstacle over `x !! y`., especially compared to writing the function that only removes one element. I don't think we have that in `lib`?

I suppose I'm not entirely opposed to returning a single module. Given my argumentation, do you still think I should do it? (or something else)

infinisil replied:
> I suppose I'm not entirely opposed to returning a single module. Given my argumentation, do you still think I should do it? (or something else)

I haven't considered that, but I really do like that idea! By returning a single module we avoid lists in the type sense while keeping the input and output types the same, that's a :+1: :+1: from me! It also aligns with the type name of `types.deferredModule` instead of `types.deferredModule`**`s`**

infinisil commented:
Nice change :D

roberth committed:
lib.types.deferredModule: Allow path-typed module references

roberth committed:
lib.types: Add deferredModuleWith

roberth committed:
lib.deferredModule: Make it properly singular

infinisil commented on lib/modules.nix:
```diff
@@ -464,7 +464,9 @@ rec {
         options = {};
-        config = addFreeformType (addMeta (removeAttrs m ["_file" "key" "disabledModules" "require" "imports" "freeformType"]));
+        config =
+          lib.throwIfNot (isAttrs m) "module ${file} (${key}) does not look like a module."
```
I think this check should be done earlier, ideally at the start of this function, since `m` is always required to be an attribute set, not just for this specific case here.

infinisil reviewed (commented):
Other than two nits this looks good. I tested this with https://github.com/infinisil/dream2nix-modules, a demo for using the module system for dream2nix created with feedback from @DavHau at Zurihac. Turns out `deferredModule` is just what was needed there

roberth committed:
lib/modules: Throw earlier when module function does not return attrs

`m` must always be an attrset at this point. It is basically always
evaluated. This will make it throw when any of the attrs is accessed,
rather than just `config`. We assume that this will improve the error
message in more scenarios.

roberth committed:
nixos/doc: Hold off on NixOps 2 info until released

roberth commented on nixos/doc/manual/development/option-types.section.md:
```diff
@@ -220,6 +220,26 @@ Value types are types that take a value parameter.
+    user to affect all submodules in an `attrsOf submodule` at once. This is
+    more convenient and discoverable than expecting the module user to
+    type-merge with the `attrsOf submodule` option. NixOps uses this type in
+    `network.defaults`.
```
I will make sure it's in 2.0, but you're right. It's too soon to mention here.

roberth commented:
> Turns out `deferredModule` is just what was needed there

Nice! I've already used it in [flake.parts `perSystem`](https://flake.parts/options.html#opt-perSystem), NixOps 2, and I will use it again in #176557

infinisil reviewed (approved):
Looking good!

roberth commented:
Thanks @infinisil!

aakropotkin commented:
@roberth do you have any usage examples? This looks useful but the only examples I could find were two mentions in a test case that didn't reveal much.

roberth commented:
[flake-parts `perSystem`](https://flake.parts/options/flake-parts.html#opt-perSystem) is one. It's for modeling anything that's a function of `system`, as used by the new Nix Flakes CLI.
